includes/derivatives.inc

```diff
@@ -474,15 +464,80 @@ function islandora_ocr_update_datastream(AbstractObject $object, $file, $dsid, $
  * Remove the generate_ocr field if present in the RELS-EXT.
  */
 function islandora_ocr_remove_generate_ocr_rels_ext_statement(AbstractObject $object) {
-  if ($object->relationships->get(ISLANDORA_RELS_EXT_URI, 'generate_ocr', '', RELS_TYPE_PLAIN_LITERAL)) {
-    $object->relationships->remove(ISLANDORA_RELS_EXT_URI, 'generate_ocr', '', RELS_TYPE_PLAIN_LITERAL);
-  }
+  module_load_include('inc', 'islandora', 'includes/utilities');
+  $message = filter_xss(islandora_deprecated('7.x-1.11', t('In the future, OCR-generating RELS-EXT statements should be removed together using islandora_ocr_remove_generating_rels_ext_statements().')));
+  trigger_error($message, E_USER_DEPRECATED);
+  islandora_ocr_remove_generating_rels_ext_statements($object, TRUE, FALSE);
 }
 
 /**
  * Set the generate_ocr relation in the RELS-EXT.
  */
 function islandora_ocr_set_generate_ocr_rels_ext_statement(AbstractObject $object, $generate_ocr = TRUE) {
+  module_load_include('inc', 'islandora', 'includes/utilities');
+  $message = filter_xss(islandora_deprecated('7.x-1.11', t('In the future, OCR-generating RELS-EXT statements should be set together using islandora_ocr_set_generating_rels_ext_statements().')));
+  trigger_error($message, E_USER_DEPRECATED);
   $generate_ocr = $generate_ocr ? 'TRUE' : 'FALSE';
-  $object->relationships->add(ISLANDORA_RELS_EXT_URI, 'generate_ocr', $generate_ocr, RELS_TYPE_PLAIN_LITERAL);
+  $object->relationships->add(ISLANDORA_RELS_EXT_URI, ISLANDORA_OCR_GENERATE_OCR_RELS_STATEMENT, $generate_ocr, RELS_TYPE_PLAIN_LITERAL);
+}
+
+/**
+ * Helper to remove one or more OCR/HOCR-generating statements.
+ *
+ * @param AbstractObject $object
+ *   Object to remove these statements from.
+ * @param bool $generate_ocr
+ *   Whether to remove the 'generate ocr' statement.
+ * @param bool $generate_hocr
+ *   Whether to remove the 'generate hocr' statement.
+ *
+ * @return array
+ *   Returns an empty array to satisfy the requirements of derivative-generating
+ *   callbacks so that this can be used as a derivative callback after
+ *   generating OCR and/or HOCR.
+ */
+function islandora_ocr_remove_generating_rels_ext_statements(AbstractObject $object, $generate_ocr = TRUE, $generate_hocr = TRUE) {
```

Comment:
Question. Lets say for N reason the one/both derivatives where no generate(failure, wrong tesseract path) can this function check somehow/get status on that before the removing the flag? Could the prdicate based flag be actually used to re-try?

Reply:
I _think_ (may be very wrong) that the intent of the API as it currently exists is to flag in all cases - if derivatives don't get generated due to some kind of error, it gets `watchdog()`ed and the flag is removed, perhaps with the intent that you're going to have to go in there again and retry anyway, at which time I don't _think_ the flag is required since the manual derivative generation passes on the `$force` flag? This way you're not setting/unsetting/setting/unsetting the flag each time you want to regenerate and making a ton of versions